pkg/rule/repo.go

```diff
@@ -0,0 +1,156 @@
+
+	Create(input rule) (rule, error)
+	GetByName(configID, name string) (rule, error)
+	UpdateWith(input rule) (rule, error)
+	ListAll(configID string) ([]rule, error)
+	ListActive(configID string, now time.Time) ([]rule, error)
+}
+
+type rule struct {
+	active      bool
+	activatedAt time.Time
+	buckets     []bucket
+	configID    string
+	createdAt   time.Time
+	criteria    *criteria
+	description string
+	deleted     bool
+	endTime     time.Time
+	id          string
+	kind        kind
+	name        string
+	startTime   time.Time
+	updatedAt   time.Time
+}
+
+type lifecycle interface {
+	setup() error
+	teardown() error
+}
+
+func (r rule) validate() (bool, error) {
+	if r.activatedAt.IsZero() {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing activatedAt attribute")
+	}
+
+	if r.buckets == nil {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing buckets attribute")
+	}
+
+	if len(r.buckets) < 1 {
+		return false, errors.Wrap(errors.ErrInvalidRule, "bucket slice contains no buckets")
+	}
+
+	if r.configID == "" {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing configID attribute")
+	}
+
+	if r.createdAt.IsZero() {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing createdAt attribute")
+	}
+
+	if r.criteria == nil {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing criteria attribute")
+	}
+
+	if r.endTime.IsZero() {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing end attribute")
+	}
+
+	if r.id == "" {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing id attribute")
+	}
+
+	if r.kind == 0 {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing kind attribute")
+	}
+
+	if r.name == "" {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing metadate.name attribute")
+	}
+
+	if r.startTime.IsZero() {
+		return false, errors.Wrap(errors.ErrInvalidRule, "missing start attribute")
+	}
+
+	totalPercentage := 0
+	for _, bucket := range r.buckets {
+		totalPercentage = totalPercentage + bucket.Percentage
+	}
+	if totalPercentage != 100 {
+		return false, errors.Wrap(errors.ErrInvalidRule, "percentage not evenly distributed")
+	}
+
+	return true, nil
+}
+
+func (r rule) run(input parameters, ctx context) (parameters, error) {
+	if r.criteria.User != nil {
+		if r.criteria.User.Age != nil {
+			return nil, errors.New("matching user age not implemented")
+		}
+
+		if r.criteria.User.ID != nil {
+			ok, err := r.criteria.User.ID.match(ctx.user.id)
+			if err != nil {
+				return nil, errors.Wrap(err, "user id match")
+			}
+
+			if !ok {
+				return nil, errors.New("rule didn't match")
+			}
+		}
+	}
+
+	params := parameters{}
+
+	switch r.kind {
+	case kindOverride:
+		params = r.buckets[0].Parameters
+	case kindExperiment:
+		return parameters{}, errors.New("experiment based rules not implemented")
+	case kindRollout:
+		return parameters{}, errors.New("rollout based rules not implemented")
+	}
+
+	for name, value := range input {
+		params[name] = value
```

Comment:
This logic seems a bit off, from how I understand it we would overwrite the params in the rule with the values from the `input` instead of the other way around. Could we check if this is doing what we expect and harden the test accordingly?